src/data_structures/fmindex.rs

```diff
@@ -101,53 +88,137 @@ impl FMIndex {
         }
 
         Interval {
+            suffix_array: self.suffix_array(),
             lower: l,
             upper: r + 1,
         }
     }
 
-    pub fn occ(&self, r: usize, a: u8) -> usize {
+
+}
+
+/// The Fast Index in Minute space (FM-Index, Ferragina and Manzini, 2000) for finding suffix array
+/// intervals matching a given pattern.
+
+#[cfg_attr(feature = "serde_macros", derive(Serialize, Deserialize))]
+pub struct FMIndex<
+        SA: Deref<Target = SuffixArray> + Clone,
+        DBWT: Deref<Target = BWT> + Clone,
+        DLess: Deref<Target = Less> + Clone,
+        DOcc: Deref<Target = Occ> + Clone> {
+    sa: SA,
```

Comment:
I would prefer if FMIndex does not need a reference to the SA, because it allows you to save memory and also serialize the FMIndex without the suffix array. Imagine you have one command line tool based on Rust-Bio that uses the FMIndex to get some suffix array intervals, and another tool that takes these intevals, loads the suffix array, and outputs text positions. This is e.g. how BWA works. If we enforce the suffix array to be present whenever we use the FMIndex, this pattern cannot be used anymore.